R/do_retention_cohort.R

```diff
@@ -1,47 +1,50 @@
-#' Run retention cohort analysis
-#' @param df Data frame to run bayes ab test
+#' Run timeseries cohort analysis
+#' @param df - Data frame to run bayes ab test
+#' @param time - time column
+#' @param value - value to aggregate
+#' @param cohort - value used to define cohort. if it is time, will be aggregated by cohort_time_unit.
+#' @param time_unit - time unit to aggregate
+#' @param fun.aggregate - aggregate function applied to values that falls under same cohort and period.
+#' @param cohort_time_unit time unit to aggregate cohort time value.
 #' @export
-do_retention_cohort <- function(df, timestamp, user_id, measure = NULL, time_unit = "month"){
+do_cohort <- function(df, time, value, cohort, time_unit = "month", fun.aggregate = n_distinct, cohort_time_unit = "month"){
   # this seems to be the new way of NSE column selection evaluation
   # ref: https://github.com/tidyverse/tidyr/blob/3b0f946d507f53afb86ea625149bbee3a00c83f6/R/spread.R
-  timestamp_col <- dplyr::select_var(names(df), !! rlang::enquo(timestamp))
-  user_id_col <- dplyr::select_var(names(df), !! rlang::enquo(user_id))
-  # the way with rlang::enquo above cannot handle NULL well. using old way.
-  measure_col <- col_name(substitute(measure))
+  time_col <- dplyr::select_var(names(df), !! rlang::enquo(time))
+  value_col <- dplyr::select_var(names(df), !! rlang::enquo(value))
+  cohort_col <- dplyr::select_var(names(df), !! rlang::enquo(cohort))
 
   grouped_col <- grouped_by(df)
 
-  # this will be executed to each group
+  # this will be executed for each group
   each_func <- function(df, ...){
-    ret <- df %>% rename_(.dots = list(.timestamp = timestamp_col))
-    if (!is.null(measure_col)) { # note that accessing measure here throws error. has to be measure_col.
-      ret <- ret %>% rename_(.dots = list(.measure = measure_col))
+    # rename columns to temporary ones first and use familiar NSE dplyr functions.
+    ret <- df %>% rename_(.dots = list(.time = time_col, .value = value_col, .cohort = cohort_col))
```

Comment:
I think you can write `rename(.time = !! time_col, .value = !! value_col, .cohort = !! cohort_col)`

It is the new way of  evaluation in tidyverse
http://dplyr.tidyverse.org/articles/programming.html